thunderbirdparser/src/org/sleuthkit/autopsy/thunderbirdparser/EmailMessageThreader.java

```diff
@@ -0,0 +1,674 @@
+        Container ref;
+
+        for (String refID : referenceList) {
+            // Check id_table to see if there is already a container for this
+            // reference id, if not create a new Container and add to table
+            ref = id_table.get(refID);
+
+            if (ref == null) {
+                ref = new Container();
+                id_table.put(refID, ref);
+            }
+
+            // Set the parent\child relationship between parent_ref and ref
+            if (parent_ref != null
+                    && !ref.hasParent()
+                    && parent_ref != ref
+                    && !parent_ref.isChild(ref)) {
+                ref.setParent(parent_ref);
+                parent_ref.addChild(ref);
+            }
+
+            parent_ref = ref;
+        }
+
+        // If the parent_ref and container are already linked, don't change 
+        // anything
+        if (parent_ref != null
+                && (parent_ref == container
+                || container.isChild(parent_ref))) {
+            parent_ref = null;
+        }
+
+        // If container already has a parent, the parent was assumed based on
+        // the list of references from another message.  parent_ref will be
+        // the real parent of container so throw away the old parent and set a
+        // new one.
+        if (container.hasParent()) {
+            container.getParent().removeChild(container);
+            container.setParent(null);
+        }
+
+        if (parent_ref != null) {
+            container.setParent(container);
+            parent_ref.addChild(container);
+        }
+    }
+
+    /**
+     * Creates a set of root container messages from the message-ID hashtable. A
+     * root Container is container that does not have a parent container.
+     *
+     * @param id_table Table of all known Containers
+     *
+     * @return A set of the root containers.
+     */
+    Set<Container> getRootSet(HashMap<?, Container> id_table) {
+        HashSet<Container> rootSet = new HashSet<>();
+
+        id_table.values().stream().filter((container)
+                -> (!container.hasParent())).forEachOrdered((container) -> {
+            rootSet.add(container);
+        });
+
+        return rootSet;
+    }
+
+    /**
+     * Remove Containers from containerSet if they do not have a message or
+     * children.
+     *
+     * @param containerSet A set of Container objects
+     */
+    void pruneEmptyContainers(Set<Container> containerSet) {
+        Set<Container> containersToRemove = new HashSet<>();
+        containerSet.forEach((container) -> {
+            if (!container.hasMessage() && !container.hasChildren()) {
+                containersToRemove.add(container);
+            } else {
+                pruneChildren(container);
+            }
+        });
+
+        containerSet.removeAll(containersToRemove);
+    }
+
+    /**
+     * Recursively work through the list of parent's children removing empy
+     * containers.
+     *
```

Comment:
I think we shoudl add to this description something to the effect of "if this passed in container does not have a message associated with it, it will get removed and its children will be assigned to its parent."

At least I think that is what is going on here.

Reply:
That is what is happening, I will put more comments.  It look me awhile to get that bit right.